pkg/leader/leader.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leader
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/rest"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ErrNoNS indicates that a namespace could not be found for the current
+// environment
+var ErrNoNS = errors.New("namespace not found for current environment")
+
+// maxBackoffInterval defines the maximum amount of time to wait between
+// attempts to become the leader.
+const maxBackoffInterval = time.Second * 16
```

Comment:
Any thoughts on making this configurable via an option?

Reply:
@hasbro17 i'd suggest we keep it a default value first. if user complains, then we make it configurable.

Reply:
Yeah I don't have any strong opinions on making it configurable in this PR. Just wanted to bring it up for discussion as something todo in the future as it might be a use case.

Reply:
I agree this could be valuable as a configuration item in the future. I concluded it wasn't a priority right now, because it's hard to come up with a compelling use case in which changing it has appreciable value. But if someone brings us a good user story, it would be easy enough to add.